Add build clarification to package guide

lwasser opened the pull request:
closes #74 

This page has been reviewed by several folks including @arianesasso @isabelizimm and was written based upon issue #74 
open for review now and will hope to merge in early september. @jagerber48 hoping this provides clarification! 

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205483903824141

NickleDave commented on images/python-package-development-process.png:
This diagram is great! It captures the key ideas we need to convey here

I'm confused about why we show documentation going into the build distribution though? None of the tree diagrams below show documentation in the sdist or wheel

lwasser replied:
oh interesting! my lens was a bit different. i was that as just a bucket of packaging "things" and then you build. not "these are the things that go into the dist" BUT your comment makes sense that most people would see that as documentation going into the build. so we can revice that graphic to just say package code, metadata (and optionally?? tests) given it's all about publishing!! and maybe i have a docs arrow that goes to a website bubble?? 

i know the test thing is really controversial and it turns out this controversy is not just in our ecosystem!!

lwasser replied:
What do you think about this? i don't want to make this graphic too complex but essentially there are some different things that could go into those SDists vs wheels. i suspect that will get confusing for newer maintainers however. 


<img width="1215" alt="Screen Shot 2023-09-25 at 3 50 51 PM" src="https://github.com/pyOpenSci/python-package-guide/assets/7649194/d94e61f2-4da5-4235-babc-21bc8b4d3042">

NickleDave commented on package-structure-code/python-package-distribution-files-sdist-wheel.md:
```diff
@@ -1,5 +1,68 @@
+
+## What is building a Python package?
+
+In Python, if you want to publish your code in a way that can be
+installed by both yourself and others, your code, tests and associated
+metadata need to be organized in a specific way. This specific
```
This first sentence might come off as a little confusing for a couple reasons:
- we have not explicitly made the link (yet, on this page) between "building" a package and "publishing" it
- the first sentence in particular might sound like it's talking about how we organize *source code*, especially if a reader doesn't already understand what is meant by metadata
- even in the next sentence, we still haven't "admitted" that when we say "specific organization and structure", we mean a *distribution package* that has been *built*--the reader doesn't find this out until the end of the paragraph.
 
I would strongly suggest we directly answer the question posed by the heading, and then say the rest. Something like this:

> To [publish your Python package](link to publishing page) and make it easy for anyone to install, you first need to *build* it. What does it mean to build a Python package? [As shown above](link to figure), when you build your Python package, you convert the source files into something called a *distribution package*. (Confusingly, the term *package* is overloaded in Python and many other languages, so on this page, we adapt the convention of the Python Packaging Authority and refer to the product of the build step as a [distribution package](https://packaging.python.org/en/latest/glossary/#term-Distribution-Package)). A distribution package contains your source code and metadata about the package, in the format required by the Python Package Index, so that it can be installed by tools like pip...

lwasser replied:
this has been fixed and pushed here @NickleDave thank you for the suggestion. I think @jagerber48 actually was getting at a similar thing where i wasn't linking building to publishing which is what a user cares about. i hope the text is better overall now.

NickleDave commented on package-structure-code/python-package-distribution-files-sdist-wheel.md:
```diff
@@ -1,5 +1,68 @@
 # The Python Package Source and Wheel Distributions
```
Commenting here because I'm not sure where else to put it -- how do we expect a reader to get to this section? Is the entire "Packaging" section meant to be read consecutively? Do we also want to cater to readers that are looking for specific information?

- If we want to cater to readers looking for specific information
  - should there be a direct link to this page from the intro page?
  - should the title be "what is building a python package?" -- if someone doesn't know the terms "source and wheel distributions" then they might not click on this page, because they won't understand why they need to know what those are
  - if the goal is for the sections to also be read consecutively, should we have a sort of "flow" on the intro page that links to each section? I feel like it's really easy to miss the "next section" links on the bottom of pages, although that's maybe my particular ADHD

lwasser replied:
excellent question @NickleDave i think because the guide has evolved, the organization may need to change. i don't want to expect anyone to read it consecutively as i know better - human behavior on websites is all about search, find what you need and keep it moving! so we want to have a navigation that supports that common use case and surface information as much as possible. in the docs section we have a bunch of individual smaller pages. we could break this up too or just as you mention surface via links. i'm really open to what you think would be best for that intended usecase (supporting people finding what they need)

lwasser replied:
ok david - i definitely renamed the page. 

`Learn about Building a Python Package`

i ive also integrated your edits!! but what i don't know is where you want links to go!! i think a flow would be good. do you mean the intro page for this specific section. and then maybe number them and explain what is on each? 

Right now, the intro page for this section looks like this:

<img width="732" alt="Screen Shot 2023-09-25 at 4 19 58 PM" src="https://github.com/pyOpenSci/python-package-guide/assets/7649194/da70c689-7c7b-42c2-abc0-0d8944f97fb3">

it would be simple to add a numbered flow.i'll try that now and show you since we're in spatial chat together and you don't know - yet - that i'm pinging you here 😆

lwasser replied:
Is this what you had in mind?? if not just let me know. i can style it more interestingly in the future as well - or maybe we move from sphinx altogether for prettier docs in the future... 

<img width="689" alt="Screen Shot 2023-09-25 at 4 27 04 PM" src="https://github.com/pyOpenSci/python-package-guide/assets/7649194/78f8e2b1-d077-488b-8912-f11199e62e65">

NickleDave reviewed (changes requested):
This looks great and is 95% of the way towards answering the question of "what is building a Python package".

I added a couple of comments about how maybe we could make the link between "build" and "distribution package" more explicit. I think we added all the info we need but it could possibly be presented more directly, so the text links in to the excellent diagram.

I also left a comment about how we will link to these sections and how we expect readers to find them. I'm afraid that might have been discussed at the meeting I missed so sorry if I'm just adding noise.

I haven't made minor typo edits, will try to come back for those later.

Will be great to have this added to the guide.

lwasser commented:
ok @NickleDave i've addressed all of your comments (i think). so let me know if you catch anything else OR if things are still not addressed correctly! i wasn't sure about the intro page comment and whether i got what you were looking for!!

jagerber48 commented on package-structure-code/python-package-distribution-files-sdist-wheel.md:
```diff
@@ -1,4 +1,72 @@
+:::{figure-md} build-workflow
+<img src="/images/python-package-development-process.png" alt="Alt tag to be added when image is final" width="700px">
+
+You need to build your Python package in order to publish it to PyPI or Conda. For PyPI, the build process essentially organizes your code and metadata into a distribution format that PyPI can read and provide to potential users.
```
Suggestions:
- Remove "essentially"
- "into a distribution format that is compatible to be uploaded to PyPI and subsequently downloaded and installed by users."

lwasser replied:
excellent - new text below. This is helpful because you're pointing out that the text is too focused on the technical nuances vs what users care about - publishing so others can use it:
******
You need to build your Python package in order to publish it to PyPI (or Conda). The build process organizes your code and metadata into a distribution format that can be uploaded to PyPI and subsequently downloaded and installed by users.

jagerber48 commented on package-structure-code/python-package-distribution-files-sdist-wheel.md:
```diff
@@ -1,4 +1,72 @@
+
+### Project metadata and PyPI
+
+For instance, when you publish to PyPI, you will notice that each package has metadata listed. Let’s have a look at [xclim](https://pypi.org/project/xclim/), one of our [pyOpenSci packages](https://www.pyopensci.org/python-packages.html). Notice that on the PyPI landing page you see some metadata about the package including python, maintainer information and more. PyPI is able to populate this me ...
```
Here there is information about how the metadata is defined using pyproject.toml and how that metadata is used to autogenerate documentation/data on the pypi page which is important. But I've gotten the sense that the metadata is also important/consumed by installers which are used for eventual installation in the user's environment. And also that it's this last fact which has led to a zoo of build tools. Maybe something about that could be mentioned here?

Maybe just a quick sentence so as not to bog down the discussion. The details about build front/backends are already discussed elsewhere I think

merwok replied:
pyproject.toml is the file for maintainers and contributors, but it should not be mentioned in a PyPI-related section.  Uploaded distributions use this standard: https://packaging.python.org/en/latest/specifications/core-metadata/

lwasser replied:
thank you both. i'm trying to digest these comments and i'm not sure how to best address them.

in this section we are trying to say that 

1. the pyproject.toml file is where a maintainer would place their metadata for their project and it needs to be formatted in a specific way to be parsed when it's built.
2. Once that build happens, for the wheel for instance that metadata is converted into another format - a METADATA file in the wheel that lives alongside other files such as a license file. 
3. It is this converted info that is uploaded to pypi in the wheel and then used by however the website / warehouse repository processes that wheel - to add metadata to the package's pypi landing page. so for instance that METATA file has classifiers, author information, etc that gets translated to the pypi website / repository. 

so while the pyproject.toml file isn't read directly by pypi the information inside of it is (indirectly) how a user populates their pypi page with correct information about their package.

pyproject.toml --> wheel / METADATA file --> pypi 

is that helpful? And if so - can we make the text more clear? 

i am going to take a leap here and @merwok please correct me if i'm off base - but i would then presume that when pip installs the package it reads that same METADATA file to grab dependency information for install purposes. but i don't know what else it might need to install. 

i see this in a METADA file for a wheel i have locally. and those are the package's dependencies. 

```
Requires-Dist: pint
Requires-Dist: pytz
Requires-Dist: arrow
Requires-Dist: requests
Requires-Dist: pydantic ==1.10.9
```

jagerber48 replied:
@lwasser 
> pyproject.toml --> wheel / METADATA file --> pypi

Yes this does help me understand what you're trying to get across here. I think I'm wanting some sentences with the gist (not the exact wording..) of:

> It is possible to supply certain pieces of package metadata in the package's `pyproject.toml` file. One use for this metadata is it will automatically appear in convenient locations for the package's PyPI page (author, classifiers) and will help make the package easier to search for so that users can find it more easily. Another important use for this metadata is supply detailed instructions about how the package should be installed (instructions about or to the installer tool??)

The second point, about metadata informing installation tools is something that is very murky to me so it's hard for me to write something intelligent about it, but from what I read, it seems like it is important to address when discussing metadata in `pyproject.toml`? An alternative view is that this section is focused on fields in the `pyproject.toml` file that are truly surface-level sort of metadata that does only really appear on the PyPI page but doesn't really go on past that. The type of information I'm talking about (I'm thinking like the `build-requires` field for example) would maybe go under a different section in this documentation, Maybe a more advanced section.

merwok replied:
This is correct, install tools use the distribution metadata (METADATA / PKG-INFO file in older versions) to find relationships between installed (or being-installed) distributions.

To add some words to the arrows:

- author writes configuration in pyproject.toml
- meta-build tool (pip or pypa-build or tox etc) reads pyproject.toml to find how to build a dist (sdist or wheel) for project
- meta-build tool invokes the build backend (which can take all its info from pyproject.toml or use other files, which can be anything – ini toml setup.cfg setup.py etc) to get a dist
- the built dist embeds METADATA, to be used by PyPI, installer tools like pip, other tools like pipdeptree or pip-audit that work with metadata for installed distributions

One complication is that sometimes a project doesn’t upload a wheel, so pip needs to take the sdist, read the pyproject in there and run all the steps above to get something it can install (to avoid having separate code paths to install wheels vs sdists).


(One last note: some projects upload wheels and no sdist, or incomplete sdist without tests or docs, which is a problem from downstream distributors. If you remember a world before the hegemony of github, projects were developed or hosted on hundreds of websites, and PyPI was the natural center point for Python distributions. Nowadays many toolchains use github repos as source, but not all. Sdists are also ill-defined (one example: should you have your cython sources or the compiled results? problems of a built source format!)

lwasser replied:
hey @merwok and @jagerber48 this is a great discussion and lots of excellent information is here. 

@merwok above you mention something that is new to me. Is tox actually a front end build tool in the same way pypa-build is? i guess i had always placed it in the nox / text runner / automation category but i also don't use tox anymore. i may have things mixed up here as well. 

in the meantime i've updated the text here and i hope it's a bit more clear - new text is below and will be pushed to this PR shortly. 


******

The metadata that both build tools and PyPI uses to describe and understand your package is generally stored in a [pyproject.toml file](pyproject-toml-python-package-metadata). This metadata is used for several purposes:

1. It helps whatever tool you use to build your package (pip, pypa's Build or an end-to-end tool such as poetry, PDM or Hatch) understand how to build your package. Information it provides to your build tool includes:

- The [build-system] table in your pyproject.toml file tells pip what [build backend tool](python-package-build-tools.html#build-back-ends) you wish to use for creating your sdist and wheel distributions.

```toml
[build-system]
requires = ["hatchling"]
build-backend = "hatchling"
```

- And the dependencies section of your project table tells the build tool and PyPI what dependencies your project requires.

```
dependencies = [
    "numpy",
    "geopandas",
]
```

2. When the build tool creates your package distribution file (the file that you publish on PyPI), it also creates a METADATA file which PyPI can read and use to help users find your package. For example:

- The `classifiers = ` section of your `[project]` table in the pyproject.toml file provides information that users on PyPI can use to filter for packages that contain specific licenses or that support specific versions of python.

```toml
classifiers = [
    # How mature is this project? Common values are
    "Development Status :: 4 - Beta",

    # Indicate who your project is intended for
    "Intended Audience :: Developers",
    "Topic :: Software Development :: Build Tools",
    "License :: OSI Approved :: MIT License",
    "Programming Language :: Python :: 3 :: Only",
    "Programming Language :: Python :: 3.10",
    "Programming Language :: Python :: 3.11",
]
```

```{admonition}
project metadata used to be stored in either a setup.py file or a setup.cfg file. The current recommended practice for storing package metadata is to use a pyproject.toml file. [Learn more about the pyproject.toml file here.](pyproject-toml-python-package-metadata)
```

### An example - xclim

When you publish to PyPI, you will notice that each package has metadata listed. Let’s have a look at [xclim](https://pypi.org/project/xclim/), one of our [pyOpenSci packages](https://www.pyopensci.org/python-packages.html). Notice that on the PyPI landing page you see some metadata about the package including python, maintainer information and more. PyPI is able to populate this metadata because it was defined using correct syntax and classifiers by Xclim's maintainers, [pyproject.toml file](https://github.com/Ouranosinc/xclim/blob/master/pyproject.toml). This metadata when the xclim package is built, is translated into a distribution file that allows PyPI to read the metadata and print it out on their website.

jagerber48 replied:
@lwasser this text is a lot more clear to me. There is a lot MORE text than there was previously, but I'm always a fan of high verbosity, so that doesn't bother me :)

lwasser replied:
i can reread it again in a day or two with fresh eyes and try to simplify a bit as well. i appreciate this feedback @jagerber48 !! 👍

merwok replied:
Yes, tox is an automation tool, primarily to run tests. It also is (or contains) a packaging build frontend, so that it can install the requirements needed by each test env.

lwasser replied:
thank you @jagerber48 
@merwok that's good to know. i didn't realize it had a packaging build frontend as well.
i am going to resolve this comment but will reread the text locally prior to publication to see if i can simplify it a bit

jagerber48 commented on package-structure-code/intro.md:
```diff
@@ -14,41 +14,62 @@ best fitted for your workflow.
+✨ 2. Learn about building your package ✨
 ^^^
-src layout, flat layout and where should tests folders live? No matter what your level of packaging knowledge is, this page will help you decide upon a package structure that follows modern python best practices.
+Building a Python package refers to the process of placing your package code
```
suggestion: 
"Building a Python package refers to the process of placing your package code and metadata into a format compatible to be published to PyPI and subsequently downloaded and installed by your users Learn more about building a Python package here."

I think I keep wanting to convert "that PyPI can read" into "that you can publish to PyPI". I don't care if PyPI can read my code, but I do care that I can publish to PyPI so that others can get my code. If that subtle distinction I'm trying to make is coming through..

lwasser replied:
got it - i'll adjust the text now. that does seem a lot more clear for user-facing documentation!! Great catch!

lwasser replied:
NEW TEXT:

Building a Python package is a great way to share your code with scientists to support open science workflow. The act of "building" refers to the process of placing your package code and
metadata into a format that can be published on PyPI. Once published on PyPI, your users can easily install it locally using pip. Learn more about building a Python package here.

merwok replied:
Shouldn’t the overall term be: Publishing a Python package is a great way etc.

In itself, building a python distribution does nothing to help other people 😉 
But publishing one (which includes building it) does it.

lwasser replied:
you are right! fixing this now!!

lwasser replied:
new text:

Publishing a Python package is a great way to share your code with scientists to support open science workflow. In order to publish a package, you will need to first build it. The act of "building" refers to the process of placing your package code and
metadata into a format that can be published on PyPI. Once published on PyPI, your users can easily install it locally using pip. Learn more about building a Python package here.

jagerber48 commented on images/python-package-development-process.png:
Nice diagram! One comment that may or may not require action: The first blue box on the left is a list of nouns. The second blue box looks like it could also be a list of nouns, or a single action, but I think it is actual a single noun "Build distribution files". Trips me up when I read it. Maybe making it "Build distribution files" (lower case d and f) and center aligning instead of left aligning would help the reader read it differently from the list of nouns in the box to the left? The bullet points already help with this.

The line for "Push SDIst, Wheel to PyPI": Should this be like `PyPI/conda/conda-forge`?

merwok replied:
No, condo is another ecosystem with its own build process and config files (that build on top of standard ones).

A general notes: the file formats should be named `sdist` and `wheel` throughout, no caps.

lwasser replied:
ok great. i've made small changes to the graphic as follows:

1. i made that initial "box" a set of files rather than a regular box to make it look different. 
2. i made sdist and wheel lowercase in the graphic and throughout this section. Oddly when you create an sdist the name SDist is actually capitalized `stravalib-1.1.0.post2-SDist.tar.gz` so i'm curious @merwok about the 
case in text vs what is actually produced by setuptools (the build for the example above). 

<img width="1446" alt="Screen Shot 2023-10-12 at 12 32 44 PM" src="https://github.com/pyOpenSci/python-package-guide/assets/7649194/93c687c4-ef9f-40b1-80cf-2120ed8ef924">

i didn't adjust the conda workflow because if they are using greyskull then you'd publish to PyPI and then add the recipe for conda-forge using greyskull. ONce that is accepted Justin, then every time you make a new release to PyPI a conda-forge build it triggered. That is how most users will publish and why i have it that way in the diagram. 

there are some instances where someone may only publish to conda for various reasons - we have one package only on bioconda channel for example. but most can co-publish - first directly to pypi and then maintain a conda repo that is triggered to update by a new release to pypi. 

please do let me know if i'm missing anything here. the new image is in this comment and also pushed locally. i'm still working on other text edits here so those aren't pushed - yet but will be today.

merwok replied:
That’s very curious, no build tool that I know does that (it’s just package name - version - .tar.gz). Do you see it in many project or just this one?

lwasser replied:
@merwok this was an anomaly . i just did a few things. 
1. i built another package locally and it did exactly what you said it should (using setuptools)
2. and i rebuilt stravalib and it also did exactly what you said it should (no sdist in the name, setuptools also). 

I am not sure what that archive was from but this is obviously a mistake on my part. perhaps i renamed it or it was created while i was testing tools. I even checked every other tagged archive on stravalib's pypi page and none have the word sdist in it. 

my apologies!!

lwasser commented:
@all-contributors please add @merwok for code, review

lwasser commented:
hey there @merwok I wanted to thank you for more great / helpful feedback on our guide. If you are interested in joining our discussions on packaging in slack, we'd welcome you there!! you can email me at leah at pyopensci.org and i can provide a link or you can provide me with your email some other way. If you have too many slacks / packaging conversations i totally understand that as well. i just wanted to extend the offer given you've been so helpful here already!

merwok commented:
I am not using slack outside of work, but feel free to tag me on github for anything where I could help!

lwasser commented:
> I am not using slack outside of work, but feel free to tag me on github for anything where I could help!

Thank you @merwok i will !! i really appreciate all of your feedback here!!

lwasser commented:
ok colleagues - i think i've addressed all of the issues / comments above at this point.

I'd love to get this merged by next week on October 31. If you have a few moments to have a look at the final changes, please do so and leave any last comments.

If not, you can ignore and things will be merged as is next week on the 31! 
If you don't have time to review but still have feedback , you can always open an issue and we can work together to get the text right / more clear together some more.

Thank you everyone for all of the feedback and again if something is still wrong / confusing i'm super open to making more changes be it this week / early next or in another PR.

lwasser commented on package-structure-code/python-package-distribution-files-sdist-wheel.md:
````diff
@@ -1,16 +1,130 @@
+```toml
+[build-system]
+requires = ["hatchling"]
+build-backend = "hatchling"
````
```suggestion
build-backend = "hatchling.build"
```

lwasser commented:
ok y'all!! i'm excited to merge this once ci finishes here! thank you all again for the input on this pr. i think once we couple this information with our tutorials we will have some great beginner-friendly / non python packaging expert friendly content!!
